images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/__init__.py

```diff
@@ -15,44 +18,94 @@
 from .actions import ActionHandler
 from .gpios import GpioManager
 
+from x1plus.utils import module_importer
+
 logger = logging.getLogger(__name__)
 
 class X1PlusDaemon:
+
     @classmethod
     async def create(cls):
         # must be a classmethod, since __init__ cannot be async
         self = X1PlusDaemon()
 
         logger.info("creating x1plusd services")
+        
+        # Required modules
         self.router = await get_dbus_router()
         self.settings = SettingsService(router=self.router)
+
         self.mqtt = MQTTClient(daemon=self)
-        self.ota = OTAService(router=self.router, daemon=self)
+        self.ota = OTAService(daemon=self)
         self.ssh = SSHService(daemon=self)
-        self.httpd = HTTPService(router=self.router, daemon=self)
-        self.sensors = SensorsService(router=self.router, daemon=self)
-        self.gpios = GpioManager(daemon=self)
-        self.expansion = ExpansionManager(router=self.router, daemon=self)
+        self.httpd = HTTPService(daemon=self)
+        self.sensors = SensorsService(daemon=self)
         self.mcproto = MCProtoParser(daemon=self)
-        self.actions = ActionHandler(router=self.router, daemon=self)
-        if not self.settings.get("polar_cloud", False):
-            self.polar_cloud = None
-        else:
-            from .polar_cloud import PolarPrintService
-            self.polar_cloud = PolarPrintService(daemon=self)
+        self.expansion = ExpansionManager(daemon=self)
+        self.gpios = GpioManager(daemon=self)
+        self.actions = ActionHandler(daemon=self)
         
+        self.custom_modules = {}
+        """
+        { "name": ClassInstance }
+        """
+
+        self.watched_keys = []
```

Comment:
the interplay between this and the shell script driver seems weird but I'm not sure how to have a better answer than that.  I wonder if a watched key should throw an exception up to the toplevel somehow or something, which would cause th erestart

Reply:
I think tossing up an exception would theoretically work, though it feels a bit like a bad design if the intent is to force an intentional restart. 

Though this would let us perhaps add the listener and function to throw the exception within `__init__.py` rather than up in `__main__`. I'm okay with either method.